### DISCUSSION TITLE
Connecting a Raspberry-Pi to the Crazyflie Through uart

### DISCUSSION THREAD

**Discussion by mandelyoni** [bitcraze/discussions#473] [Q&A]
Dear Madam/Sir,

Please advise the best options to **connect and communicate** between a **Crazyflie** and a **Raspberry**-Pi .

Thanks,  Yoni

**Comment by krichardsson** [comment]
Hi, 

There are a few options :-)

If you want wireless communication, the Crazyradio and the pythonlib should work fine. We use this in out test-lab.

For wired connections you can:
1. Use USB + python lib
2. Use UART + python lib using CRTP, see [documentation](https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/0.1.21/development/uart_communication/)
3. Use UART + your own code to talk to the UART. More work, but you can implement whatever protocol you want
4. Use one of the other busses on the expansion port such as SPI or similar. Even more work...

1 and 2 are pretty straight forward. I'm not sure about the performance impact, but my guess is that the UART solution might be better for the CF as the USB code runs at a very high priority, but I'm guessing. We usually don't fly with the USB cable connected :-)
1 and 2 also enables you to use all the infrastructure in the python lib.
If you use the python lib you will have to write some sort of python app on the pi as well.

Kristoffer

**Comment by mandelyoni** [comment]
Thanks Kristoffer, we'll try it out.

**Reply by mandelyoni** [reply]
Hi Kristoffer,

It is not working when connecting through wires. 
We followed:  https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/0.1.21/development/uart_communication/
Attached are the .config file, the RaspberryPi's terminal response and the physical connection.

Thanks,  Yoni
![RaspberryPi response](https://user-images
![bolt_PI](https://user-images.githubusercontent.com/4273822/209466200-3296cff1-f760-4ff8-a472-f644f5df1509.jpeg)
.github
[config_github.txt](https://github.com/bitcraze/discussions/files/10299821/config_github.txt)
usercontent.com/4273822/209466157-3a6437c7-05e6-40aa-83a1-6833594ab16a.jpeg)

**Reply by mandelyoni** [reply]
![RaspberryPi response](https://user-images.githubusercontent.com/4273822/209466285-22235288-9782-4403-9dcb-0414747ce636.jpeg)

**Comment by knmcguire** [comment]
It seems that there has been an bug in the crazyflie python library that causes this serial error. We just merged this PR to fix that: https://github.com/bitcraze/crazyflie-lib-python/pull/373

Could you install the cflib from source via these instructions on the raspberry pi? https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/0.1.21/installation/install/

**Comment by zlite** [comment]
A test report after trying the above with the new cflib compiled from source, which does work. But there are a few gotchas/bugs:

1) It appears it doesn't close the port correctly on the CF side. The first time the example works, but it hangs on the second time. You have to power cycle the CF. 

2) Using the UART immediately kills the Crazyradio stream. You can't use them together. 

Regarding using the USB connection, is there documentation on that? I'm connecting CF to a RPi and would love to use the USB rather than the UART if possible

**Comment by zlite** [comment]
Although the CF shows up as ttyACM0 on the Pi when plugged into the serial port, it won't connect using the demo script modified to use that port. 

CF is not recognized on a PC via USB despite the drivers being loaded

**Reply by knmcguire** [reply]
Hi! For the 'CF is not recognized on a PC via USB despite the drivers being loaded': could you make a new discussion thread in Q&A for this, with more info like OS and such? 

For the raspberry pi, we will try this out shortly to see if we can recreate it

**Reply by knmcguire** [reply]
I've changed the script for testing it, as it seems that the console script was a bit unstable. But even so, it was always connecting, at least after powercycling the crazyflie. 

The most recent guide information is to be found here. https://github.com/bitcraze/crazyflie-lib-python/blob/master/docs/development/uart_communication.md

**Reply by zlite** [reply]
Just to be clear, the observed issues are these:

1) It connects fine over UART, but needs to be powercycled before each connection. That connection also kills the Crazyradio feed
2) It won't connect over USB, despite changing to ttyACM0, which is what the CF shows up as when plugged into a RPi

**Reply by knmcguire** [reply]
Ah like that, USB goes over a different protocol. Have you tried `usb://0` ? If you set the usb permissions via these instructions, that should work: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/installation/usb_permissions/

**Reply by zlite** [reply]
CORRECTION!  I just had to reboot the Pi and now it works over USB! Apparently reloading the rules was not enough. 

Thanks again for your help

**Reply by zlite** [reply]
Also, over USB it doesn't require a restart to run the script again, which is good news

**Reply by zlite** [reply]
Connecting over USB does kill the CFRadio feed, however, just like connecting over UART

**Reply by knmcguire** [reply]
Yes I guess that is by design. CRTP is the protocol that is used for these connections and that only works through one channel by design. Perhaps if you want to use CRTP through radio, then perhaps the uart communication needs to be set up more manually. That requires a bit of hacking but I would say that it would be doable with CPX:
- Documentation: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/cpx/
- Example script of the AIdeck FPV shows this I think? https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/aideck/fpv.py
Might not be the best example but perhaps you get the picture.

**Reply by zlite** [reply]
Ah, got it. I'm more used to MAVlink, which can handle many channels at once

**Reply by knmcguire** [reply]
Yeah.. we know. Currently, we are in the process of redesigning the comms stack with the soon arrival of Crazyradio 2 (see [this blogpost](https://www.bitcraze.io/2022/11/new-crazyradio-and-new-communication-architecture/)), and enabling multiple links is one of them. 

I was told by @ataffanel is that there is a hardswitch in the firmware that allows only one link the time for CRTP, but this could be removed by hacking the firmware a bit... ofcourse since it is not designed for multiple links who knows what will happen, but would be a good test 😄 ?

If you want to know how to do this I would advise to start a new discussion though, since that is going quite offtopic of this one.
